### PR TITLE
classic frogger game done

### PR DESCRIPTION
**Pull request by andysmokk**
### Object-Oriented JavaScript (Classic Frogger Game)

[Demo](https://andysmokk.github.io/classic-frogger-game/) |
[Code base](https://github.com/andysmokk/classic-frogger-game)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by andysmokk** [comment]
Self-check complete. Thanks:v:.

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  if (this.x > 500) {
+    this.x = INITIAL_POSITION_ENEMIES_X;
+    this.speed = this.randomInteger();
+  }
+
+  if (
+    player.y < this.y + 50 &&
+    player.y + 50 > this.y &&
+    player.x < this.x + 75 &&
+    player.x + 75 > this.x
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:32]: Please find a requirement among those listed/linked by bot, post it here as a comment and also fix code accordingly.

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
+// Enemies our player must avoid
+const INITIAL_POSITION_PLAYER_X = 203;
+const INITIAL_POSITION_ENEMIES_X = -200;
+const INITIAL_POSITION_ENEMY_ONE_Y = 62;
+const INITIAL_POSITION_ENEMY_TWO_Y = 145;
+const INITIAL_POSITION_ENEMY_THREE_Y = 228;
+const INITIAL_POSITION_PLAYER_Y = 390;
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:7]: Think of less wordy contants definition.
Requirements provide relevant examples.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@andysmokk good job!
Yet one important thing needs to be fixed. Constants definition is secondary and less critical but fix it as well.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 2 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js]: Classes do not refer to any global variables, like global variable player, which is an instance of Player class (referring to global constants and globals provided by the gaming platform like Resources is OK); Hint: pass instance of a game object (or objects) as an argument to other game objects they need to interact with.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 1 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js]: Classes do not refer to any global variables, like global variable player, which is an instance of Player class (referring to global constants and globals provided by the gaming platform like Resources is OK); Hint: pass instance of a game object (or objects) as an argument to other game objects they need to interact with.

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  return Math.floor(MIN_SPEED + Math.random() * (MAX_SPEED + 1 - MIN_SPEED));
+};
+
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+const enemyOne = new Enemy(POSITION_ENEMIES_X, POSITION_ENEMY_ONE_Y);
+const enemyTwo = new Enemy(POSITION_ENEMIES_X, POSITION_ENEMY_TWO_Y);
+const enemyThree = new Enemy(POSITION_ENEMIES_X, POSITION_ENEMY_THREE_Y);
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:48]: Imagine you are tasked to add 10 more enemies.
Make this fragment DRY and number of line of codes not dependent on number of enemies.

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+var Player = function (x, y) {
+  this.x = x;
+  this.y = y;
+  this.sprite = "images/char-boy.png";
+};
+
+Player.prototype.initialPositionPlayer = function () {
+  player.x = POSITION_PLAYER_X;
+  player.y = POSITION_PLAYER_Y;
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:60]: Explain, why `player.*`?

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+  if (
+    gamer.y < this.y + 50 &&
+    gamer.y + 50 > this.y &&
+    gamer.x < this.x + 75 &&
+    gamer.x + 75 > this.x
+  ) {
+    gamer.initialPositionPlayer();
+  }
+};
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:36]: This shouldn't work unless you changed the `engine`.
Please re-do as changing the `engine` is a no-go here.
Imagine, you are on a commercial project and opt to change shared code when the problem is solvable without doing this.
The requirements contain a hint on how to resolve the issue properly. Just need to read carefully.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@andysmokk good job.
Although needs improvement.
You may want to revisit OOP educational materials.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 4 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js:60]: Бо я не уважний . У цьому випадку щоб отримати доступ до даних всередині об'єкта треба використовувати this, а не глобальну змінну. Виправлено.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 5 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js]: Fixed.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 3 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js]: Fixed.

**Comment by andysmokk** [comment]
Дуже дякую Вам, я дійсно слабкий в ООП.
Ваші підказки та вказання на помилки дали мені можливість розібратись та зрозуміти ООП краще.

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  this.y = y;
+  this.user = user;
+  this.speed = this.randomInteger();
+  this.sprite = "images/enemy-bug.png";
+};
+
+Enemy.prototype.update = function (dt) {
+  this.x += this.speed * dt;
+
+  if (this.x > 500) {
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:21]: What is 500?

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    this.x = POSITION_ENEMIES_X;
+    this.speed = this.randomInteger();
+  }
+
+  if (
+    this.user.y < this.y + 50 &&
+    this.user.y + 50 > this.y &&
+    this.user.x < this.x + 75 &&
+    this.user.x + 75 > this.x
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:30]: What are 50 and 75?

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+var Player = function (x, y) {
+  this.x = x;
+  this.y = y;
+  this.sprite = "images/char-boy.png";
+};
+
+const player = new Player(POSITION_PLAYER_X, POSITION_PLAYER_Y);
+
+const enemyPositionsY = [62, 145, 228];
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:52]: What are 62, 145, 228?
As a peer deveoper I am tasked to add the 4th track enemies can run. What would be the number? Can we calculate it based on track number?

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+Player.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function (keyPressed) {
+  if (keyPressed === "up" && this.y > 0) {
+    this.y -= BLOCK_HEIGHT;
+  }
+
+  if (keyPressed === "down" && this.y < 390) {
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:77]: What is 390?
Is it calculable?

File: submissions/andysmokk/classic-frogger-game/app.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    this.y += BLOCK_HEIGHT;
+  }
+
+  if (keyPressed === "left" && this.x > 1) {
+    this.x -= BLOCK_WIDTH;
+  }
+
+  if (keyPressed === "right" && this.x < 405) {
```
Comment (OleksiyRudenko) [inline submissions/andysmokk/classic-frogger-game/app.js:85]: What is 405?
Is it calculable?

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@andysmokk good job.
Let's polish it.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 6 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js:21]: Is the wigth of the playing field. Fixed.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 7 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js:30]: Fixed.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 8 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js:52]: Fixed. Added a calculate of function the position Y of enemies.

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 9 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js:77]: 390 is the player`s starting point. It is calculable. Fixed

File: submissions/andysmokk/classic-frogger-game/app.js (hunk 10 above)
Comment (andysmokk) [reply submissions/andysmokk/classic-frogger-game/app.js:85]: 405 is the end point of the player (axis X). It is calculable. Fixed

**Comment by andysmokk** [comment]
Please review again 😏.

**Comment by andysmokk** [comment]
Thank you.

**Pull request merged**